pilot/pkg/xds/sds.go

```diff
@@ -0,0 +1,291 @@
+			}
+			rootCert = extractRoot(k8sSecret)
+		}
+	}
+	if rootCert == nil {
+		log.Warnf("could not find ca certificate for %v/%v", res.Namespace, strippedName)
+	}
+
+	return &tls.Secret{
+		Name: res.ResourceName,
+		Type: &tls.Secret_ValidationContext{
+			ValidationContext: &tls.CertificateValidationContext{
+				TrustedCa: &core.DataSource{
+					Specifier: &core.DataSource_InlineBytes{
+						InlineBytes: rootCert,
+					},
+				},
+			},
+		},
+	}
+}
+
+func getKeyAndCert(secrets informersv1.SecretInformer, res SecretResource) *tls.Secret {
+	k8sSecret, err := secrets.Lister().Secrets(res.Namespace).Get(res.Name)
+	if errors.IsNotFound(err) {
+		adsLog.Warnf("requested secret %v/%v not found", res.Namespace, res.Name)
+		return nil
+	}
+	if err != nil {
+		adsLog.Errorf("failed to get secret %v/%v: %v", res.Namespace, res.Name, err)
+		return nil
+	}
+
+	cert, key := extractCertAndKey(k8sSecret)
+	if key == nil || cert == nil {
+		adsLog.Warnf("could not extract secret from %v/%v, key and cert must be present", res.Namespace, res.Name)
+		return nil
+	}
+	return &tls.Secret{
+		Name: res.ResourceName,
+		Type: &tls.Secret_TlsCertificate{
+			TlsCertificate: &tls.TlsCertificate{
+				CertificateChain: &core.DataSource{
+					Specifier: &core.DataSource_InlineBytes{
+						InlineBytes: cert,
+					},
+				},
+				PrivateKey: &core.DataSource{
+					Specifier: &core.DataSource_InlineBytes{
+						InlineBytes: key,
+					},
+				},
+			},
+		},
+	}
+}
+
+// extractCertAndKey extracts server key, certificate
+func extractCertAndKey(scrt *v1.Secret) (cert, key []byte) {
+	if len(scrt.Data[genericScrtCert]) > 0 {
+		cert = scrt.Data[genericScrtCert]
+		key = scrt.Data[genericScrtKey]
+	} else {
+		cert = scrt.Data[tlsScrtCert]
+		key = scrt.Data[tlsScrtKey]
+	}
+	return cert, key
+}
+
+// extractRoot extracts the root certificate
+func extractRoot(scrt *v1.Secret) (cert []byte) {
+	if len(scrt.Data[genericScrtCaCert]) > 0 {
+		return scrt.Data[genericScrtCaCert]
+	} else if len(scrt.Data[tlsScrtCaCert]) > 0 {
+		return scrt.Data[tlsScrtCaCert]
+	}
+	return nil
+}
+
+type SecretGen struct {
+	secrets informersv1.SecretInformer
+	queue   queue.Instance
+}
+
+var _ model.XdsResourceGenerator = &SecretGen{}
+
+func NewSecretGen(k kube.Client, updater model.XDSUpdater) *SecretGen {
+	secrets := k.KubeInformer().Core().V1().Secrets()
+	secrets.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				handle(obj, updater)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				handle(cur, updater)
+			},
+			DeleteFunc: func(obj interface{}) {
+				handle(obj, updater)
+			},
+		})
+	return &SecretGen{
+		secrets: k.KubeInformer().Core().V1().Secrets(),
+	}
+}
+
+func handle(obj interface{}, updater model.XDSUpdater) {
+	scrt, ok := obj.(*v1.Secret)
+	if !ok {
+		adsLog.Errorf("Failed to convert to secret object: %v", obj)
+		return
+	}
+	updater.ConfigUpdate(&model.PushRequest{
+		Full: true,
```

Comment:
May be add a comment here that if the secrets change rapidly, then we may need to re-eval this. For now, its okay as we dont expect gateway secrets to change often.

Reply:
Just pushed an update to do incremental pushes